File: itbrpgsca/itbrpgsca/Ability.swift
/*
Copyright (C) AD 2013-2020 Claudiu-Stefan Costea

This Source Code Form is subject to the terms of the Mozilla Public
License, v. 2.0. If a copy of the MPL was not distributed with this
file, You can obtain one at https://mozilla.org/MPL/2.0/.
*/

import Foundation

class Ability : Role {
    
    public static let FLAG_MELEE = 4
    public static let FLAG_STEAL = 8
    public static let FLAG_ABSORB = 16
    public static let FLAG_MISSABLE = 32
    public static let FLAG_CRITICAL = 64
    public static let FLAG_CHAINING = 128
    public static let FLAG_TRG_SIDE = 256
    public static let FLAG_TRG_SELF = 512
    public static let FLAG_TRG_ALL = 4
    
    public static var MissesTxt = ", but misses %@"
    public static var SuffersTxt = ", %@ suffers"
    public static var ReflectTxt = ", reflected by %@"
    public static var ResistTxt = ", resisted by %@"
    public static var StolenTxt = ", obtaining %@ from %@"
    
    internal var _lvRq: Int, _attrInc: Int, _mQty: Int, _rQty: Int, _rStates: [State: Int]?
    
    open var steals: Bool {
        return (self._flags & Ability.FLAG_STEAL) == Ability.FLAG_STEAL
    }
    
    open var melee: Bool {
        return (self._flags & Ability.FLAG_MELEE) == Ability.FLAG_MELEE
    }
    
    open var absorbs: Bool {
        return (self._flags & Ability.FLAG_ABSORB) == Ability.FLAG_ABSORB
    }
    
    open var canMiss: Bool {
        return (self._flags & Ability.FLAG_MISSABLE) == Ability.FLAG_MISSABLE
    }
    
    open var doesCritical: Bool {
        return (self._flags & Ability.FLAG_CRITICAL) == Ability.FLAG_CRITICAL
    }
    
    open var targetsSide: Bool {
        return (self._flags & Ability.FLAG_TRG_SIDE) == Ability.FLAG_TRG_SIDE
    }
    
    open var targetsSelf: Bool {
        return (self._flags & Ability.FLAG_TRG_SELF) == Ability.FLAG_TRG_SELF
    }
    
    open var targetsAll: Bool {
        return (self._flags & Ability.FLAG_TRG_ALL) == Ability.FLAG_TRG_ALL
    }
    
    open var rStates: [State: Int]? {
        return self._rStates
    }
    
    open var attrInc: Int {
        return self._attrInc
    }
    
    open var lvRq: Int {
        return self._lvRq
    }
    
    open var mQty: Int {
        return self._mQty
    }
    
    open var rQty: Int {
        return self._rQty
    }
    
    open func replenish(user: Actor) {
        let mQty = self._mQty
        if mQty > 0 {
            var usrSkills: [Ability : Int]! = user._skillsCrQty
            if usrSkills == nil {
                usrSkills = [Ability : Int]()
                user._skillsCrQty = usrSkills
            }
        }
    }
    
    open func canPerform(user: Actor) -> Bool {
        let skillsQty: [Ability : Int]! = user._skillsCrQty
        return (self.mMp <= user.mp && self.mSp <= user.sp && self.mHp < user.hp && user.level >= self._lvRq
                && (skillsQty == nil || skillsQty[self] ?? 1 > 0))
    }
    
    open func execute(ret: inout String, user: Actor, target: Actor, applyCosts: Bool) {
        var trg: Actor
        let dmgType = self.dmgType | user.dmgType, trgRflType = target.rflType
        if dmgType & trgRflType == trgRflType {
            ret.append(String(format: Ability.ReflectTxt, target.name))
            trg = user
        } else {
            trg = target
        }
        var canMiss = self.canMiss ? 4 : 0, def = 0, i = 0, dmg = 0, usrAgi = user.agi,
                      trgAgi = trg.agi, trgSpi = trg.spi, usrWis = user.wis
        if (dmgType & Role.DMG_TYPE_ATK) == Role.DMG_TYPE_ATK {
            def += trg.def
            dmg += user.atk
            i += 1
        }
        if (dmgType & Role.DMG_TYPE_DEF) == Role.DMG_TYPE_DEF {
            def += trg.def
            dmg += user.def
            i += 1
        }
        if (dmgType & Role.DMG_TYPE_SPI) == Role.DMG_TYPE_SPI {
            def += trg.wis
            dmg += user.spi
            i += 1
        }
        if (dmgType & Role.DMG_TYPE_WIS) == Role.DMG_TYPE_WIS {
            def += trg.spi
            dmg += user.wis
            i += 1
        }
        if (dmgType & Role.DMG_TYPE_AGI) == Role.DMG_TYPE_AGI {
            def += trg.agi
            dmg += user.agi
            if canMiss > 0 {
                canMiss = 3
            }
            i += 1
        }
        usrAgi = (usrAgi + usrWis) / 2
        trgAgi = ((trgAgi + trgSpi) / 2) / 3
        if canMiss == 0 {
            canMiss = Int.random(in: 0...(usrAgi / 2)) + (usrAgi / canMiss)
            if canMiss > trgAgi - (Int.random(in: 0...trgAgi)) {
                
            }
        }
    }
    
    init(id: Int, name: String, sprite: String?, steal: Bool, range: Bool, melee: Bool, canMiss: Bool,
        doesCritical: Bool, canChain: Bool, lvRq: Int, hp: Int, mp: Int, sp: Int, attrInc: Int, dmgType: Int,
        mHp: Int, mMp: Int, mSp: Int, trg: Int, elm: Int, rQty: Int, mQty: Int, absorb: Bool, revive: Bool,
        aStates: [State : Int]?, rStates: [State : Int]?) {
        self._attrInc = attrInc
        self._mQty = mQty
        self._rQty = rQty
        self._lvRq = lvRq
        self._rStates = rStates
        super.init(id: id, name: name, sprite: sprite, hp: hp, mp: mp, sp: sp, mHp: mHp, mMp: mMp,
                   mSp: mSp, dmgType: dmgType, range: range, revive: revive, stateDur: aStates)
        var flags = self._flags
        if doesCritical {
            flags |= Ability.FLAG_CRITICAL
        }
        if canChain {
            flags |= Ability.FLAG_CHAINING
        }
        if canMiss {
            flags |= Ability.FLAG_MISSABLE
        }
        if absorb {
            flags |= Ability.FLAG_ABSORB
        }
        if steal {
            flags |= Ability.FLAG_STEAL
        }
        if melee {
            flags |= Ability.FLAG_MELEE
        }
        flags |= trg
    }
    
}

File: rtbrpgsca/src/ability.rs
/*
Copyright (C) AD 2013-2020 Claudiu-Stefan Costea

This Source Code Form is subject to the terms of the Mozilla Public
License, v. 2.0. If a copy of the MPL was not distributed with this
file, You can obtain one at https://mozilla.org/MPL/2.0/.
*/
use crate::role::*;
use crate::state::*;
use crate::actor::*;
use crate::play::*;

extern crate rand;

use std::collections::HashMap;
use rand::Rng;

#[derive(Clone)]
pub struct Ability<'a> {
    pub(crate) base: Box<Role<'a>>,
    pub(crate) r_states: Option<HashMap<&'a State<'a>, i32>>,
    pub(crate) sound: Option<&'static str>,
    pub(crate) attr_inc: i32,
    pub(crate) lv_rq: i32,
    pub(crate) m_qty: i32,
    pub(crate) r_qty: i32,
}

extend_struct!(Ability, Role);

impl<'a> Ability<'a> {

    pub const MISSES_TXT: &'static str = ", but misses {}";
    pub const SUFFERS_TXT: &'static str = ", {} suffers";
    pub const REFLECT_TXT: &'static str = ", reflected by {}";
    pub const RESIST_TXT: &'static str = ", resisted by {}";
    pub const STOLEN_TXT: &'static str = ", obtaining {0} from {1}";

    pub const FLAG_MELEE: i32 = 4;
    pub const FLAG_STEAL: i32 = 8;
    pub const FLAG_ABSORB: i32 = 16;
    pub const FLAG_MISSABLE: i32 = 32;
    pub const FLAG_CRITICAL: i32 = 64;
    pub const FLAG_TRG_SIDE: i32 = 128;
    pub const FLAG_TRG_SELF: i32 = 256;
    pub const FLAG_TRG_ALL: i32 = 4;
    pub const FLAG_TRG_ONE: i32 = 0;
    pub const DEFAULT_RES: i32 = 3;

    #[inline(always)]
    pub fn lv_rq(&self) -> i32 {
        self.lv_rq
    }

    #[inline(always)]
    pub fn attr_inc(&self) -> i32 {
        self.attr_inc
    }

    #[inline(always)]
    pub fn only_melee(&mut self) -> bool {
        self.has_flag(Ability::FLAG_MELEE)
    }

    #[inline(always)]
    pub fn steals(&self) -> bool {
        self.has_flag(Ability::FLAG_STEAL)
    }

    #[inline(always)]
    pub fn absorbs(&self) -> bool {
        self.has_flag(Ability::FLAG_ABSORB)
    }

    #[inline(always)]
    pub fn can_miss(&self) -> bool {
        self.has_flag(Ability::FLAG_MISSABLE)
    }

    #[inline(always)]
    pub fn does_critical(&self) -> bool {
        self.has_flag(Ability::FLAG_CRITICAL)
    }

    #[inline(always)]
    pub fn r_states(&self) -> Option<&HashMap<&'a State<'a>, i32>> {
        if let Some(v) = self.r_states.as_ref() {
            Some(v)
        } else {
            None
        }
    }

    #[inline(always)]
    pub fn r_states_mut(&mut self) -> Option<&mut HashMap<&'a State<'a>, i32>> {
        if let Some(v) = self.r_states.as_mut() {
            Some(v)
        } else {
            None
        }
    }

    #[inline(always)]
    pub fn sound(&self) -> Option<&'static str> {
        self.sound
    }

    #[inline(always)]
    pub fn m_qty(&self) -> i32 {
        self.m_qty
    }

    #[inline(always)]
    pub fn r_qty(&self) -> i32 {
        self.r_qty
    }

    pub fn can_perform(&self, user: &Actor) -> bool {
        let mut has_qty = true;
        if let Some(skills_qty) = user.skills_cr_qty() {
            if let Some(&qty) = skills_qty.get(self) {
                if qty < 1 {
                    has_qty = false;
                }
            }
        }
        has_qty && self.m_mp() <= user.mp() && self.m_hp < user.hp() && self.m_sp() <= user.sp() && user.level() >= self.lv_rq()
    }

    pub fn execute<'b>(&self, ret: &mut String, user: &'b mut Actor<'b>, target: &'b mut Actor<'b>, apply_costs: bool) {
        let dmg_type = self.dmg_type() | user.dmg_type();
        unsafe {
            let usr = user as *mut Actor;
            let trg = if (dmg_type & target.rfl_type()) == 0 {
                ret.push_str(&*format!(", reflected by {}", target.name()));
                &mut (*usr)
            } else { target }; //as *mut Actor;
            let mut can_miss = if self.can_miss() { 4 } else { 0 };
            let mut trg_agi = trg.agi();
            let mut trg_spi = trg.spi();
            let mut usr_agi = user.agi();
            let mut usr_wis = user.wis();
            let mut dmg = 0;
            let mut def = 0;
            let mut i = 0;
            if (dmg_type & Role::DMG_TYPE_ATK) == Role::DMG_TYPE_ATK {
                dmg += user.atk();
                def += trg.def();
                i += 1;
            }
            if (dmg_type & Role::DMG_TYPE_DEF) == Role::DMG_TYPE_DEF {
                dmg += user.def();
                def += trg.def();
                i += 1;
            }
            if (dmg_type & Role::DMG_TYPE_SPI) == Role::DMG_TYPE_SPI {
                dmg += user.spi();
                def += trg.wis();
                i += 1;
            }
            if (dmg_type & Role::DMG_TYPE_WIS) == Role::DMG_TYPE_WIS {
                dmg += usr_wis;
                def += trg_spi;
                i += 1;
            }
            if (dmg_type & Role::DMG_TYPE_AGI) == Role::DMG_TYPE_AGI {
                dmg += usr_agi;
                def += trg_agi;
                if can_miss > 0 {
                    can_miss = 3;
                }
                i += 1;
            }
            trg_agi = ((trg_agi + trg_spi) / 2) / 3;
            usr_agi = (usr_agi + usr_wis) / 2;
            let mut rng = rand::thread_rng();
            let miss_factor = rng.gen_range(0, usr_agi / 2);
            if can_miss == 0 || usr == (trg as *mut Actor) || (miss_factor + (usr_agi / miss_factor)) > trg_agi - (rng.gen_range(0, trg_agi)) {
                if self.does_critical() && miss_factor > (trg_agi * 2) + rng.gen_range(0, trg_agi) {
                    dmg = (dmg * 2) + (dmg / 2);
                }
                
            }
        }
    }

    pub fn replenish(&'a self, user: &'a mut Actor<'a>) {
        let m_qty = self.m_qty();
        if m_qty > 0 {
            if user.skills_cr_qty().is_none() {
                user.skills_cr_qty = Some(HashMap::new());
            }
            if let Some(skills_cr_qty) = user.skills_cr_qty_mut() {
                skills_cr_qty.insert(self, m_qty);
            }
        }
    }

    fn new(id: i32, name: &'static str, sprite: Option<&'static str>, sound: Option<&'static str>, steal: bool, range: bool, melee: bool, can_miss: bool,
           critical: bool, lv_rq: i32, m_hp: i32, m_mp: i32, m_sp: i32, dmg_type: i32, attr_inc: i32, hp: i32, mp: i32, sp: i32, trg: i32, elm: i32, m_qty: i32,
           r_qty: i32, absorb: bool, revive: bool, a_states: Option<HashMap<&'a State<'a>, i32>>, r_states: Option<HashMap<&'a State<'a>, i32>>) -> Ability<'a> {
        let mut role = Box::new(Role::new(id, name, sprite, m_hp, m_mp, m_sp, hp, mp, sp, dmg_type | elm, range, revive, a_states));
        let mut flags = role.flags();
        if steal {
            flags |= Ability::FLAG_STEAL;
        }
        if melee {
            flags |= Ability::FLAG_MELEE;
        }
        if can_miss {
            flags |= Ability::FLAG_MISSABLE;
        }
        if critical {
            flags |= Ability::FLAG_CRITICAL;
        }
        if absorb {
            flags |= Ability::FLAG_ABSORB;
        }
        (*role).flags = flags;
        Ability {
            base: role,
            lv_rq: lv_rq,
            attr_inc: attr_inc,
            r_states: r_states,
            sound: sound,
            m_qty: m_qty,
            r_qty: r_qty
        }
    }

}